### PR TITLE
refactor(plugin-server): refactor the event pipeline

### PR DESCRIPTION
**Pull request by macobo**
## Problem

This PR refactors how the event pipeline works. It's a follow-up to #9738 

The problem being solved by these PRs is that the [event processing pipeline](https://mermaid-js.github.io/mermaid-live-editor/edit#pako:eNqlVltP2zAU_iuWERJItNVeq2kSg03joWMaSHtY9mDsk8Zqame-FCHEf58TJ61rO4FteWnq853vHJ9rnjGVDPASF6Ks5SOtiDLo_roQyD1AaLUCrckabsQatOFS_CzwlnCx4MPB7LcFC3qxIeWG-D9zo09yugX-5Yn9yacdCPNulNAfzKBFOca98jclqSPutEeVGw861vb6w6-2D2tFmgo1XFMuiD9tnx9SbUCtwFSShdbyiOAyecCtmBRe0i424iAPGD30IOoOde-UVHnvUhCzFLrTe_nRluUE45UCYiAirIhgNXhHV8TQCvRBWHLBLgX7zBV8kXITSBSsuTagAsXwmmEYtL-zdy6yzlNRYMOK4yCEwn2Ko0seEG-O6V_H9bXYHiN9hG-YE_Hy6VZd1pzoYwgljbEKosOO8s7Ruwh_ByoVc-GKfAfB-oZuX_zrlRTabkG5DqL969zF0vUJms0-ZFt_fCh0KmFTo7TP0fsWlAY3PPGYsfZzrj6GJTNvQrJDg4_pe_aoYPaki6a2ay70wgHaARYX1oF_ol66QEzkckq18-4oya-jp7LvOU5P0VdpAJmKmL6TPAHXqAFVSrUF5gSoIYrUNdTokZsKyXBmxWlU4HwUGpGur9HWTwQU5y5QS1IXeJLdCglZP0QTosHRgeRNVH4gZbi8AJ3xMrrceZ__tMICU3EbeJgz40trpkHtXKFpRRd9ze331ui2S-h85idK42324gV5kjRT3nre-MgczDpfYPAFyC4Q1LBtuQqc4U1WUd6bzHLqqyAnaTXyG2sMndliLsSZU3Smyc51mkTs4TyYF8OojQdrGKnkOKna3IYcGxNJOtBsnred1nNqp3MlWcL_VNYn6S4fSi2R_F-t8OwtMjnGF9jlxn1FMvcV_NwqF9hUzlSBl-6VQUls7dwsxIuD2oa1dhg3UuFlSWoNF5hYI--eBMVLoywMoGtO3MfHtke9_AF0KwJj) was really hard to follow due to confusing terminology, worker/main thread split, return value dependencies and more.

This in turn is needed to split plugin-server processing in a nice way.

## Changes

Event pipeline is now a pipeline of steps where every step can call a subsequent one or stop processing. 

Note that the problem _isn't_ fully solved - `process-event.ts` probably does too much right now, but we're in a better position than before.

Also note that this way of structuring things also lent itself really well for unit testing and caught some bugs that previously slipped through.

### CODE REVIEW

File: plugin-server/src/worker/ingestion/event-pipeline/determineShouldBufferStep.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import { DateTime } from 'luxon'
+
+import { Hub, Person, PreIngestionEvent, TeamId } from '../../../types'
+import { EventPipelineRunner, StepResult } from './runner'
+
+export async function determineShouldBufferStep(
```
[the remark this answers is not in the document]
Comment (yakkomajuri) [reply plugin-server/src/worker/ingestion/event-pipeline/determineShouldBufferStep.ts:6]: Also I just realized the naming is also problematic because `determine` implies to me that this has no side effects - it just makes a decision (and returns a boolean or something). Instead it actually sends an event to the buffer.

More like `sendEventToBufferIfNeeded` or something

File: plugin-server/src/worker/ingestion/event-pipeline/runAsyncHandlersStep.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+): Promise<StepResult> {
+    const promises = []
+    let actionMatches: Action[] = []
+    if (event.event !== '$snapshot') {
+        actionMatches = await runner.hub.actionMatcher.match(event, person, elements)
+        promises.push(runner.hub.hookCannon.findAndFireHooks(event, person, event.siteUrl, actionMatches))
+    }
+
+    const processedPluginEvent = convertToProcessedPluginEvent(event)
+    const isSnapshot = event.event === '$snapshot'
```
Comment (yakkomajuri) [inline plugin-server/src/worker/ingestion/event-pipeline/runAsyncHandlersStep.ts:21]: we could move this up given we have a `isSnapshot` test on line 15

**Review by yakkomajuri — CHANGES_REQUESTED** [review]
Awesome! This monster was not as bad as it seemed given all the tests. 

Very close to approving ✅ 

Left some comments but didn't find any major gaps, love the test coverage. Let's maybe get this in on Monday though and I'll make sure to also keep an eye out

File: plugin-server/src/worker/ingestion/event-pipeline/runAsyncHandlersStep.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+export async function runAsyncHandlersStep(
+    runner: EventPipelineRunner,
+    event: PreIngestionEvent,
+    person: Person | undefined,
+    elements: Element[] | undefined
+): Promise<StepResult> {
+    const promises = []
+    let actionMatches: Action[] = []
+    if (event.event !== '$snapshot') {
+        actionMatches = await runner.hub.actionMatcher.match(event, person, elements)
```
[the remark this answers is not in the document]
Comment (macobo) [reply plugin-server/src/worker/ingestion/event-pipeline/runAsyncHandlersStep.ts:16]: I'm not sure what you mean. Do you want to ignore errors from action matching when deciding whether to run onAction? That's a new requirement if so.

File: plugin-server/src/worker/ingestion/event-pipeline/runAsyncHandlersStep.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                func: (event) => runOnAction(runner.hub, actionMatch, event),
+                statsKey: `kafka_queue.on_action`,
+                timeoutMessage: 'After 30 seconds still running onAction',
+            })
+        )
+    }
+
+    await Promise.all(promises)
+
+    return null
```
Comment (yakkomajuri) [inline plugin-server/src/worker/ingestion/event-pipeline/runAsyncHandlersStep.ts:46]: Just an ordering change to guarantee `onEvent` runs if e.g. `runner.hub.actionMatcher.match` fails

```suggestion
    const promises = []

    const processedPluginEvent = convertToProcessedPluginEvent(event)
    const isSnapshot = event.event === '$snapshot'
    const method = isSnapshot ? runOnSnapshot : runOnEvent
    promises.push(
        runInstrumentedFunction({
            server: runner.hub,
            event: processedPluginEvent,
            func: (event) => method(runner.hub, event),
            statsKey: `kafka_queue.single_${isSnapshot ? 'on_snapshot' : 'on_event'}`,
            timeoutMessage: `After 30 seconds still running ${isSnapshot ? 'onSnapshot' : 'onEvent'}`,
        })
    )

    let actionMatches: Action[] = []
    if (!isSnapshot) {
        actionMatches = await runner.hub.actionMatcher.match(event, person, elements)
        promises.push(runner.hub.hookCannon.findAndFireHooks(event, person, event.siteUrl, actionMatches))
    }

    for (const actionMatch of actionMatches) {
        promises.push(
            runInstrumentedFunction({
                server: runner.hub,
                event: processedPluginEvent,
                func: (event) => runOnAction(runner.hub, actionMatch, event),
                statsKey: `kafka_queue.on_action`,
                timeoutMessage: 'After 30 seconds still running onAction',
            })
        )
    }

    await Promise.all(promises)

    return null
```

File: plugin-server/src/worker/ingestion/event-pipeline/runAsyncHandlersStep.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (yakkomajuri) [reply plugin-server/src/worker/ingestion/event-pipeline/runAsyncHandlersStep.ts:16]: No - let me submit a suggestion as to what I want

File: plugin-server/tests/postgres/teardown.test.ts [hunk 5]
```diff
@@ -111,13 +120,15 @@ describe('teardown', () => {
             [pluginConfig39.id],
             'testTag'
         )
-        const event1 = await piscina!.run({ task: 'processEvent', args: { event: { ...defaultEvent } } })
+        const event1 = await processEvent(piscina, defaultEvent)
+        // const event1 = await piscina!.run({ task: 'runEventPipeline', args: { event: { ...defaultEvent } } })
```
Comment (Twixes) [inline plugin-server/tests/postgres/teardown.test.ts:124]: Do we still need this? Commented out code is generally confusing

File: plugin-server/src/worker/ingestion/event-pipeline/runner.ts [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    private async runPipeline<Step extends StepType, ArgsType extends StepParameters<EventPipelineStepsType[Step]>>(
+        name: Step,
+        ...args: ArgsType
+    ): Promise<EventPipelineResult> {
+        let currentStepName: StepType = name
+        let currentArgs: any = args
+
+        while (true) {
```
[the remark this answers is not in the document]
Comment (macobo) [reply plugin-server/src/worker/ingestion/event-pipeline/runner.ts:84]: That's more confusing IMO as that's literally impossible.

File: plugin-server/src/worker/ingestion/event-pipeline/runAsyncHandlersStep.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (macobo) [reply plugin-server/src/worker/ingestion/event-pipeline/runAsyncHandlersStep.ts:16]: I mistyped my question - onEvent calling should not be affected by action-related errors? If so, I think re-ordering is too implicit about that and we should make that obvious in the code.

File: plugin-server/src/worker/ingestion/event-pipeline/prepareEventStep.ts [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        team_id,
+        DateTime.fromISO(now),
+        sent_at ? DateTime.fromISO(sent_at) : null,
+        uuid!, // it will throw if it's undefined,
+        site_url
+    )
+
+    if (preIngestionEvent && preIngestionEvent.event !== '$snapshot') {
+        return runner.nextStep('determineShouldBufferStep', preIngestionEvent)
+    } else if (preIngestionEvent && preIngestionEvent.event === '$snapshot') {
```
[the remark this answers is not in the document]
Comment (macobo) [reply plugin-server/src/worker/ingestion/event-pipeline/prepareEventStep.ts:22]: Intentional - I was trying to make it clearer to the reader it's dealing with a snapshot here.

File: plugin-server/src/worker/ingestion/event-pipeline/runAsyncHandlersStep.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (macobo) [reply plugin-server/src/worker/ingestion/event-pipeline/runAsyncHandlersStep.ts:16]: Let's resolve this in a follow-up PR - I'm in merge conflict hell until this is in.

File: plugin-server/src/worker/ingestion/event-pipeline/prepareEventStep.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (yakkomajuri) [reply plugin-server/src/worker/ingestion/event-pipeline/prepareEventStep.ts:22]: fair

**Review by yakkomajuri — APPROVED** [review]
